docarray/index/backends/epsilla.py

```diff
@@ -0,0 +1,527 @@
+                if getattr(self, attr, None) is not None
+            ]
+
+            if missing_attributes:
+                raise ValueError(
+                    f"Missing required attributes for self-hosted version: {', '.join(missing_attributes)}"
+                )
+
+        def validate_cloud_config(self):
+            missing_attributes_cloud = [
+                attr
+                for attr in ["cloud_project_id", "cloud_db_id", "api_key"]
+                if getattr(self, attr, None) is not None
+            ]
+
+            if missing_attributes_cloud:
+                raise ValueError(
+                    f"Missing required attributes for cloud version: {', '.join(missing_attributes_cloud)}"
+                )
+
+    @dataclass
+    class RuntimeConfig(BaseDocIndex.RuntimeConfig):
+        # No dynamic config used
+        pass
+
+    @property
+    def collection_name(self):
+        return self._db_config.table_name
+
+    @property
+    def index_name(self):
+        return self.collection_name
+
+    def python_type_to_db_type(self, python_type: Type) -> str:
+        # AbstractTensor does not have n_dims, which is required by Epsilla
+        # Use NdArray instead
+        for allowed_type in [list, np.ndarray, AbstractTensor]:
+            if safe_issubclass(python_type, allowed_type):
+                return 'VECTOR_FLOAT'
+
+        py_type_map = {
+            ID: 'STRING',
+            str: 'STRING',
+            bytes: 'STRING',
+            int: 'BIGINT',
+            float: 'FLOAT',
+            bool: 'BOOL',
+            np.ndarray: 'VECTOR_FLOAT',
+        }
+
+        for py_type, epsilla_type in py_type_map.items():
+            if safe_issubclass(python_type, py_type):
+                return epsilla_type
+
+        raise ValueError(f'Unsupported column type for {type(self)}: {python_type}')
+
+    def _index(self, column_to_data: Dict[str, Generator[Any, None, None]]):
+        self._index_subindex(column_to_data)
+
+        rows = list(self._transpose_col_value_dict(column_to_data))
+        normalized_rows = []
+        for row in rows:
+            normalized_row = {}
+            for key, value in row.items():
+                if isinstance(value, NdArray):
+                    normalized_row[key] = value.tolist()
+                elif isinstance(value, np.ndarray):
+                    normalized_row[key] = value.tolist()
+                else:
+                    normalized_row[key] = value
+            normalized_rows.append(normalized_row)
+
+        status_code, response = self._db.insert(
+            table_name=self._table_name, records=normalized_rows
+        )
+
+        if status_code != HTTPStatus.OK:
+            raise IOError(
+                f"Failed to insert documents. "
+                f"Error code: {status_code}. Error message: {response}."
+            )
+
+    def num_docs(self) -> int:
+        raise NotImplementedError
```

Comment:
There is currently no public API for counting the number of docs. One workaround is to run `filter` query without any condition, but I worry that is quite inefficient. Would you prefer to implement `num_docs` with that rather than throwing here?

Reply:
I double checked. `num_docs` impacts these two methods:
- `__len__` 
- `_is_index_empty`

I pushed a `_filter` based implementation for `_is_index_empty`. Confirmed there is nothing that depends on `num_docs` is limited (nothing in `abstract.py` depends on this). But there is no easy way to check all the `len()` locations.

```
$ grep -rn -l --exclude-dir=./tests "num_docs()"
./docarray/index/abstract.py
./docarray/index/backends/milvus.py
./docarray/index/backends/redis.py
./docarray/index/backends/hnswlib.py
./docs/user_guide/storing/index_hnswlib.md
./docs/user_guide/storing/index_elastic.md
./docs/user_guide/storing/index_in_memory.md
./docs/user_guide/storing/index_qdrant.md
./docs/user_guide/storing/index_redis.md
```

Reply:
and _is_index_empty, I think it was checked in many queries. I am not sure if the result is kind of cached. I have little access to check properly now the code